### PR TITLE
null-text-inversion-pipeline-implementation

### PR DESCRIPTION
**Pull request by Junsheng121**
# What does this PR do?
Issue: [https://github.com/huggingface/diffusers/issues/6313](url)
The implementation of a Null-Text Inversion Pipeline. NullTextPipeline is mostly from the official code [[https://github.com/google/prompt-to-prompt/blob/main/null_text_w_ptp.ipynb](url)

Usage:
```
from diffusers.schedulers import DDIMScheduler
from pipeline_null_text_inversion import NullTextPipeline
import torch

invert_prompt = "A lying cat"
input_image = "siamese.jpg"
steps = 50
prompt = "A lying cat"

# must torch.float32
scheduler = DDIMScheduler(num_train_timesteps=1000, beta_start=0.00085, beta_end=0.0120, beta_schedule="scaled_linear")
pipeline = NullTextPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", scheduler = scheduler, torch_dtype=torch.float32).to("cuda")

inverted_latent, uncond = pipeline.invert(input_image, invert_prompt, num_inner_steps=5, early_stop_epsilon= 1e-5, num_inference_steps = steps)
pipeline(prompt, uncond, inverted_latent, guidance_scale=7.5, num_inference_steps=steps).images[0].save(input_image+".output.jpg")
```

invert_prompt is used for DDIM-Inversion and optimization.

prompt can be the same with invert_prompt for reconstruction, or be set to a different one like "A lying dog" for image editing. 

Note that float16 will fail to successfully optimize the null-text embedding.

Fixes # (issue)

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Could you please:

* Remove the image from the PR? 
* Run the styling check (`make style && make quality`)? 
* Add a section about the pipeline in the README?

**Review by patrickvonplaten — COMMENTED** [review]
Could we remove the image `siamese.jpg` from the PR to make sure the repo is kept light-weight and add some lines to the README?

**Comment by Junsheng121** [comment]
OK

Patrick von Platen ***@***.***>于2023年12月27日 周三05:27写道：

**Comment by patrickvonplaten** [comment]
Now we just need to run `make style` once and we should be good :-)

**Comment by Junsheng121** [comment]
Done~

**Comment by patrickvonplaten** [comment]
> Done~

Perfect!

**Comment by Skquark** [comment]
While implementing my UI for Null-Text, ran into a few issues worth pointing out.. In the doc, it shows to import with ```from examples.community.pipeline_null_text_inversion import NullTextPipeline``` but took me a while to figure out that it should be ```from pipeline_null_text_inversion import NullTextPipeline``` instead. Then the scheduler in example gave warnings with steps_offset & clip_sample, so it should be ```scheduler = DDIMScheduler(num_train_timesteps=1000, beta_start=0.00085, beta_end=0.0120, beta_schedule="scaled_linear", steps_offset=1, clip_sample=False)```... Not sure if it's compatible with all the other schedulers, but I can stick with DDIM for now. Then I found it strange that the input image is supposed to be a path string instead of a PIL or Numpy like every other pipeline, but I could live with that.  Also wouldn't mind a callback function for my progress bar, seems like generation time is slow but probably worth it.
The main issue when I finally got it working is the returned StableDiffusionPipelineOutput which gave error about .save not being on str.  I checked the returned images value and it was a list with a string, so ```["images"]``` was the output data.  I looked into the script, and at the very end I saw this line 256 ```image, output_type=output_type, do_denormalize=[True] * image.shape[0]``` which looks like the variables are reversed, so the variable image = output_type, which is "images". Then on the last line is ```return StableDiffusionPipelineOutput(images=image, nsfw_content_detected=False)``` where I think images should be a list, but maybe it can take a single.  Everything looks like easy fixes, looking forward to testing the results, seems useful for edits.  Thanks..

**Comment by Junsheng121** [comment]
I will fix it later

Alan Bedian ***@***.***> 于2024年1月17日周三 15:20写道：

**Comment by mycfhs** [comment]
Hello! Is there any way to use it with float16?
As you say,  float16 will fail.  I find the reason may be the grad turn to zero in unet layers due to it's float16 dtype. The grad in unet is about 1e-10 ~ 1e-7,  and those who less than 1e-8 may turn to zero due to the effective range of float16. And following layers' grad turn to smaller and smaller, and finally to zero.

I tried to enlarge the loss by a factor of 10(or 1e3, 1e4), or lr to smaller, but all failed. Are there any solutions to solve it? I really really want to use it with torch.float16 o(╥﹏╥)o

**Comment by Adenialzz** [comment]
Hi, thanks for contributing. How to combine this with prompt-to-prompt for image editing? Could you please provide an example code snippet? Thanks.
